### PR TITLE
Feature/tile unit factory units

### PR DESCRIPTION
**Pull request by karthshen**

### CODE REVIEW

File: src/main/game/gameboard/Tile.java [hunk 1]
```diff
@@ -2,203 +2,108 @@
...
-//        containsArmy = true;
 //    }
-//
-//    public void removeArmy(int armyID) {
-//        for (int i = 0; i < armies.size(); i++) {
-//            if (armies.get(i).getArmyID() == armyID) {
+
+    //test if terrain is impassable
+    public boolean isImpassable(){
+        return (Terrain == TerrainEnum.INVISIBLE || Terrain == TerrainEnum.WATER) ? true : false;
```
Comment (randygbrooks) [inline src/main/game/gameboard/Tile.java]: The  ? true : false; is pointless, fix this.

File: src/main/game/gameboard/Tile.java [hunk 2]
```diff
@@ -2,203 +2,108 @@
...
-//
-//        return false;
-//    }
+    }
+
+    public boolean containsStructure(){
+        if(structure!=null)
+            return true;
+        else
+            return false;
```
Comment (randygbrooks) [inline src/main/game/gameboard/Tile.java]: This is similar to the other one. Why not just return (structure != null)

File: src/main/game/gameboard/Tile.java [hunk 3]
```diff
@@ -2,203 +2,108 @@
...
 //    }
-//
-//    public void removeArmy(int armyID) {
-//        for (int i = 0; i < armies.size(); i++) {
-//            if (armies.get(i).getArmyID() == armyID) {
+
+    //test if terrain is impassable
+    public boolean isImpassable(){
+        return (Terrain == TerrainEnum.INVISIBLE || Terrain == TerrainEnum.WATER) ? true : false;
+    }
```
Comment (randygbrooks) [inline src/main/game/gameboard/Tile.java]: The ? true : false; is pointless, fix this.

File: src/main/game/gameboard/Tile.java [hunk 4]
```diff
@@ -2,203 +2,108 @@
...
+        if(structure!=null)
+            return true;
+        else
+            return false;
+    }
 
+    public boolean containsUnit(){
+        return false;
+    }
+
```
Comment (randygbrooks) [inline src/main/game/gameboard/Tile.java]: Why do these just return false?

**Comment by gdscheele** [comment]
don't forget to squash commits before you merge. Slack me if you forget how

File: src/main/game/entities/EntityId.java [hunk 5]
```diff
@@ -28,4 +28,13 @@ public Enum getSubTypeId(){
     public int getInstanceId(){
         return this.instanceId;
     }
+
+    //If entityId equals, return 1, else return 0;
+    public int compareTo(EntityId entityId){
+        if(this.playerId == entityId.getPlayerId()&&this.instanceId==entityId.getInstanceId()
```
Comment (gdscheele) [inline src/main/game/entities/EntityId.java]: can you make the formatting on the && consistent?

Comment (gdscheele) [inline src/main/game/entities/EntityId.java]: Also, you don't need an else statement if you're just returning in both cases, but that's up to you

File: src/main/game/entities/EntitySubtypeEnum.java [hunk 6]
```diff
@@ -0,0 +1,5 @@
+package game.entities;
+
+public enum EntitySubtypeEnum {
+    MELEE, RANGE, COLONIST, EXPLORER, CAPITOL, FARM, MINE, POWERPLANT, OBSERVATIONTOWER, FORT, UNIVERSITY;
```
Comment (gdscheele) [inline src/main/game/entities/EntitySubtypeEnum.java]: Can you change it to OBSERVATION_TOWER?

File: src/main/game/entities/factories/UnitFactory.java [hunk 7]
```diff
@@ -11,28 +15,64 @@
...
-            case COLONIST:
-                return new Colonist(unitStatistics.get(unit), location);
-            case EXPLORER:
-                return new Explorer(unitStatistics.get(unit), location);
+            case COLONIST:{
+                try{
+                    newColonistId = this.colonistIdManager.getNewId();
+                    entityId = new EntityId(playerId, EntityTypeEnum.UNIT, EntitySubtypeEnum.COLONIST, newColonistId );
+                    return new Colonist(unitStatistics.get(unit), location, entityId);
+                } catch (IdLimitExceededException e) {}
```
Comment (gdscheele) [inline src/main/game/entities/factories/UnitFactory.java]: Handle these exceptions somehow. Probably want to throw something like ColonistLimitExceeded, ExplorerLimitExceeded, etc. for the caller to handle.

File: src/main/game/gameboard/Tile.java [hunk 8]
```diff
@@ -2,203 +2,108 @@
...
+    public Location getLocation(){return location;}
+
+    public void addUnit(Unit unit){
+        units.add(unit);
+        //unit.setLocation(this.location);
+    }
+
+    public void addStructure(Structure structure){
+        if(containsStructure()==false){
+            this.structure = structure;
```
Comment (gdscheele) [inline src/main/game/gameboard/Tile.java]: should set this.containsStructure to true after you add a structure

Comment (karthshen) [inline src/main/game/gameboard/Tile.java]: If you look at the containsStructure(), the boolean is determined base on if the structure == null, thus, once the structure is created the method will automatically return false;

**Pull request closed without merging**